fetch_submissions.py

```diff
@@ -140,14 +143,21 @@ def save_submission_as_file(submission):
     outfile.write(submission['code'])
 
 
+def get_user_credentials():
+  for i in range(5):
```

Comment:
Instead of having a for loop here which makes it hard to understand why it is there, you could have something like this:

```python
def get_user_credentials(retry_count=0):
```

This way, you could recursively call the function with a one less retry count when a username or password is wrong. And it makes the code a bit more easier to understand.

Next, I am not sure if having print statements within a function whose only purpose is to get the user logged in. I feel like logic to prompt the user and get input belongs in `__main__`

Reply:
For the first point, I agree that the "magic constant" 5 shouldn't be there. Instead of making the function recursive, I can just use a variable such as `max_retry_attempts` to make the intent clear. 

For the second one, I personally think it's better to have a separate function for login as the `__main__` looks cleaner. Returning an error message on unsuccessful login is a part of the whole login process so I think that it makes sense to have a print statement within that function.

Reply:
Sure you can change the name of the variable to be clearer. I just happen to find the recursive function more elegant. Also, it would allow the caller of the `login` function to specify how many times would you want to retry which I feel is better than having a local variable inside `login` or a global variable at the top of the file.

As for the second part, I feel like the `login` function's only job is to try logging in or if it fails, inform the caller that it failed. The caller should handle what to do with it. I find that to be a much clearer separation.

Reply:
> As for the second part, I feel like the login function's only job is to try logging in or if it fails, inform the caller that it failed. The caller should handle what to do with it. I find that to be a much clearer separation.

What if I create a `LoginError` custom exception which is thrown by the `get_user_credentials` functions when `retry_count == 0` condition is met? This exception can then be handled in a try-catch block within `__main__`.

Reply:
I'm thinking of getting rid of the `get_user_credentials` function altogether. I'll let the `__main__` handle the error message and retry login if unsuccessful.